### PR TITLE
Add CMake as a supported build system

### LINKED ISSUE #50: Re-enable ccache in CI
When/If #46 is merged, CI on this repo can no longer take advantage of ccache because the cache directory wouldn't be saved to GHA cahce nor recovered. Ccache should be re-enabled with a volume/bind to the docker container and https://github.com/actions/cache.

### PR DESCRIPTION
**Pull request by lunacd**
Closes #50. Supersedes #78.

## Changes

This commit adds CMake as a supported build system alongside meson.

## Benefits

For a potentially low-level project like cps-config that many other C++
projects would have a reason to depend on, using CMake instead of meson
removes python as a transitive dependency for those C++ projects that
choose to leverage cps-config.

CMake is also just a very widely used build system for C++. Providing
CMake as an option may introduce a more familiar workflow for those that
would like to use, contribute to, or package cps-config.

## Future work

This introduces a fair amount of repetition between the cmake and meson
GitHub workflow files. Ideally this can be extracted into an action that
the two can share. I'll leave figuring out exactly how that works for
another day.

---

*This PR also refactors the CI process a bit to introduce the flexibility necessary for this change.*

# Perform build outside docker build process

## Changes

Currently, source code copied into docker images and the built binaries
are also stored there.

This commit extracts that process from docker build into the CI
pipeline.

## Benefits

- Docker image layers are cached. Moving source code and binaries out of
docker image reduces unnecessary cache, which are never going to be a
cache-hit for CI purposes anyway.
- This also adds flexibility to support other build systems other than
meson.

### CODE REVIEW

File: CMakeLists.txt
Comment (dcbaker) [inline CMakeLists.txt:5]: I'd prefer not to hardcode the version separately in every supported build system, I opened #83, which adds a `VERSION` file, I think you can use `file(READ ...)` to accomplish the same basic thing in CMake?

File: src/CMakeLists.txt
Comment (dcbaker) [inline src/CMakeLists.txt:5]: This is different than Meson, which has C++17 (I picked C++17 because I wanted to have broader compatibility), but whatever we are using we should use the same standard for all build systems unless there's a really compelling reason not to.

File: src/CMakeLists.txt
Comment (dcbaker) [inline src/CMakeLists.txt:45]: Is there a reason not to do a version check?

File: tests/docker/rockylinux.Dockerfile
Comment (dcbaker) [inline tests/docker/rockylinux.Dockerfile:23]: These docker changes are also a separate logical change, can those go in a third commit?

File: tests/CMakeLists.txt
Comment (dcbaker) [inline tests/CMakeLists.txt:7]: I think it's at least noting with `TODO` style comments anything that the meson build files are able to do that the CMake isn't (like running the python based tests).

File: src/CMakeLists.txt
Comment (dcbaker) [inline src/CMakeLists.txt:24]: This feels like a lot of repetition, CMake doesn't have a builtin mechanism to turn a bool into an int, or even a ternary?

**Review by dcbaker — CHANGES_REQUESTED** [review]
I don't have a problem with including CMake support, I understand that a lot of people either have not choice or prefer CMake, and unfortunately the duplication is unavoidable. I have a few comments on things, and I opened #83 to try to ease a little bit of the duplication.

I would note that I have gone out of my way to make sure that the Meson works with Muon, which is a pure C99 implementation of Meson, so there is another option for people who are concerned about Python.

File: .clang-format
Comment (lunacd) [reply .clang-format:194]: I should have better explained why this is here.

Adding CMake introduces `CMakePresets.json`. Without this change CI will fail because `clang-format` will try to format json file without knowing how. The alternative change I could make here is to make `clang-format` not format JSON files. I'm nots sure how but I'm sure there's a way.

File: CMakeLists.txt
Comment (lunacd) [reply CMakeLists.txt:5]: Makes total sense. I'll rebase and refactor once #83 is merged.

File: src/CMakeLists.txt
Comment (lunacd) [reply src/CMakeLists.txt:5]: Ooopsie. That should have never been there. I was testing whether `std::unreachable` detection is working as expected.

File: src/CMakeLists.txt
Comment (lunacd) [reply src/CMakeLists.txt:45]: Good point. I'll add version checks consistent with meson

File: tests/docker/rockylinux.Dockerfile
Comment (lunacd) [reply tests/docker/rockylinux.Dockerfile:23]: Sure. I was trying to make every commit have a passing CI, but I agree separating those will be clearer

File: src/CMakeLists.txt
Comment (lunacd) [reply src/CMakeLists.txt:24]: The only one I know of is https://cmake.org/cmake/help/latest/command/configure_file.html#transformations. I'm not sure how to make it work with an dynamic number of macros.

File: tests/CMakeLists.txt
Comment (lunacd) [reply tests/CMakeLists.txt:7]: This is an unintentional omission. I'll add it

File: .gitignore
Comment (lunacd) [reply .gitignore:7]: I feel like, someone more familiar with CMake please correct me if I'm wrong, this has been the common way of ignoring build directories with CMake. CMake doesn't produce a `.gitignore` in its build directory, which I think is because of support for in-source build. I've seen people hacking to make CMake do what meson does, see [this](https://www.scivision.dev/cmake-auto-gitignore-build-dir/), but I personally think that's uglier than adding it to gitignore.

File: .clang-format
Comment (lunacd) [reply .clang-format:194]: I can separate this to a separate commit explaining why tho.

File: tests/CMakeLists.txt
Comment (lunacd) [reply tests/CMakeLists.txt:7]: Help wanted: how do I make cmake consume TAP output? Current attempt that does not properly consume TAP output:
```cmake
# Integration tests
find_package(
  Python 3.11
  COMPONENTS Interpreter
  REQUIRED)
add_test(
  NAME pkg-config-compatibility-tests
  COMMAND
    ${CMAKE_COMMAND} -E env CPS_PATH="${CMAKE_CURRENT_SOURCE_DIR}/cases"
    ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/runner.py
    $<TARGET_FILE:cps-config> ${CMAKE_CURRENT_SOURCE_DIR}/cases.toml)

```

**Comment by lunacd** [comment]
@dcbaker Thanks a lot for your really helpful reviews! I have fixed some of them and will get back to the others soon

File: CMakeLists.txt
Comment (dcbaker) [reply CMakeLists.txt:5]: Merged :)

File: src/CMakeLists.txt
Comment (dcbaker) [reply src/CMakeLists.txt:22]: Google suggests that maybe something like `$<IF:${HAS_BUILTIN_${function_upper}},1,0>`, but maybe some of the more CMake fluent people have suggestions.

Comment (lunacd) [reply src/CMakeLists.txt:22]: Oh sorry I misunderstood your original comment. I was thinking of something else. Getting rid of the if else block

File: tests/CMakeLists.txt
Comment (lunacd) [reply tests/CMakeLists.txt:7]: @dcbaker Google seems to indicate that CMake does not have a consumer for TAP tests. I'm reading from the TAP spec that it explicitly disregard exit code? Would it break anything if I make `runner.py` return non-zero on failure?

Comment (lunacd) [reply tests/CMakeLists.txt:8]: I implemented this change. Let me know if this breaks anything. If it does I'll look for another solution.

Comment (dcbaker) [reply tests/CMakeLists.txt:8]: It should be fine as long as it returns 0 for success and 1 for failure. TAP does have some return code stuff, but it's not documented outside of perl code that I could find. I'm also open to switching away from TAP, it was just pretty easy to implement and Meson could consume it easily

Comment (lunacd) [reply tests/CMakeLists.txt]: I'm honestly not sure what protocol CMake supports best. It seems to me that CMake expects a different command being run for each test case. This would require either hard-coding a list of testcases in CMakeLists or have first party support for a testing framework through something like `gtest_discover_tests`. Or of course treat the entire suite as one test case. None of the three sounds optimal to me.

I think having it return 1 on failure is the route that makes most sense currently.

**Comment by lunacd** [comment]
@dcbaker I'm not quite sure I have understood [this part](https://github.com/cps-org/cps-config/blob/cc40978ecf88e8ca194881ba333323f734143f95/src/cps/meson.build#L4-L8) correctly.

The way I'm reading it, it's checking for the existence of `unreachable` in the standard library, i.e. `std::unreachable`. But I'm not sure why the macro is guarding [`__builtin_unreachable`](https://github.com/cps-org/cps-config/blob/cc40978ecf88e8ca194881ba333323f734143f95/src/cps/utils.hpp#L28-L31). I'm also not sure why meson successfully detects `std::unreachable` when compiling in C++17. That's why I think I might have misunderstood something.

I have made the CMake version check for `std::unreachable` but I may have been trying the wrong thing in the first place.

**Comment by lunacd** [comment]
@bretbrownjr Would appreciate review for some CMake best practices :)

**Comment by dcbaker** [comment]
@lunacd Meson's `has_function` has some unfortunate legacy behavior where `has_function` will work with regular functions, function-like-macros (I think), and builtins that have known prefixes. We've (Meson upstream) been discussing changing that, but trying to nail down what exactly the future will look like has been slow. 

In this case we're looking for `__builtin_unreachable` not `std::unreachable`. @tylerjw has pointed out that the design I'm using `__builtin_unreachable` for has some problems anyway, and I think we've agreed we're going to refactor it out eventually anyway.

**Comment by lunacd** [comment]
Oh I see. Thanks for explaining how our current meson setup works! I checked docs all the way back to gcc 4 and clang 4. They both support `__builtin_unreachable`. Since they are currently the two compilers we are targeting, let me just hard-code that macro to 1 for now. When we refactor it out, we could revisit how that part of CMake works.

**Comment by dcbaker** [comment]
Since clang also has had support for a long time, I think that's a fine solution.

File: .gitignore
Comment (bretbrownjr) [reply .gitignore:7]: I personally see ignoring `build/` as slightly more clutter than needed, but ultimately benign clutter. I don't feel strongly either, but I think I'd like to see PRs moving along.

File: CMakeLists.txt
Comment (bretbrownjr) [inline CMakeLists.txt:3]: Using a tool like [bumpversion](https://pypi.org/project/bumpversion/) can keep this kind of special wiring logic to a minimum, but no need to hold up this PR on that. Someone can circle back to that in the future if it's interesting.

File: CMakeLists.txt
Comment (bretbrownjr) [inline CMakeLists.txt:13]:
```suggestion
    enable_testing()
```

`include(CTest)` implies `enable_testing()`.

File: src/CMakeLists.txt
Comment (bretbrownjr) [inline src/CMakeLists.txt:3]:
```suggestion
add_library(libcps
  cps/env.cpp
  cps/loader.cpp
  cps/printer.cpp
  cps/search.cpp
  cps/utils.cpp
  cps/version.cpp
  )
```
Not a big deal, but putting each file on a different source line _generally_ makes it easier to trivially rebase multiple PRs that add or remove files. Trivial rebasing is really nice when trying to get PRs to land from the GitHub web UI.

File: src/CMakeLists.txt
Comment (bretbrownjr) [inline src/CMakeLists.txt:36]:
```suggestion
find_package(tl-expected REQUIRED)
target_link_libraries(libcps PRIVATE tl::expected)

find_package(fmt REQUIRED)
target_link_libraries(libcps PRIVATE fmt::fmt)

find_package(jsoncpp REQUIRED)
target_link_libraries(libcps PRIVATE JsonCpp::JsonCpp)

find_package(CLI11 REQUIRED)
target_link_libraries(cps-config PRIVATE CLI11::CLI11)
```

Unless we have a specific problem we're trying to work around, it's best to not put specific version information in a build configuration. That info really goes in dependency management logic, for instance in the ubuntu build environment setup logic.

File: src/CMakeLists.txt
Comment (bretbrownjr) [reply src/CMakeLists.txt:3]: Also, I'd just call the library `cps` and not `libcps`, but I don't think I can make a simple PR suggestion to clean that up.

File: src/CMakeLists.txt
Comment (bretbrownjr) [reply src/CMakeLists.txt:36]: I would also move the `add_executable` down to line 34 or so. It's good to keep all the declaration and configuration for a target all together. It's good to have all of the context for how a target is configured all on the same "page" of code, or as close to that as you can get. Just for clarity reasons.

File: src/cps/config.hpp.in
Comment (bretbrownjr) [inline src/cps/config.hpp.in:3]: I'm not a huge fan of putting version information in header files as it pessimizes build caching in general, though keeping feature parity across CMake and meson is more important for this PR to land. A followon PR to consider another approach would be interesting to me. Eventually.

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:1]:
```suggestion
enable_testing()
```

This line is harmless but unneeded. It's generally best to just put `include(CTest)` in the top-level `CMakeLists.txt` to describe the whole project as being "test capable". It's a property of the project as a whole.

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:2]:
```suggestion
find_package(GTest REQUIRED)
```

See above about avoiding overspecifying versions (Do we care if we find a compatible googletest > 1.10? Probably not.).

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:5]:
```suggestion
add_executable(cps-tests
  loader.cpp
  utils.cpp
  version.cpp
)
```

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:7]:
```suggestion
target_link_libraries(cps-tests PRIVATE
  GTest::gtest
  GTest::gtest_main
)
```

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:16]:
```suggestion
  REQUIRED
)
```

Just for consistent paren placement.

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:23]:
```suggestion
    $<TARGET_FILE:cps-config> ${CMAKE_CURRENT_SOURCE_DIR}/cases.toml
)
```

File: tests/CMakeLists.txt
Comment (bretbrownjr) [inline tests/CMakeLists.txt:21]: You can also set properties on tests to adjust the environment variables for the tests, but this implementation works too.

See:

* https://cmake.org/cmake/help/latest/command/set_tests_properties.html
* https://cmake.org/cmake/help/latest/prop_test/ENVIRONMENT_MODIFICATION.html

**Review by bretbrownjr — CHANGES_REQUESTED** [review]
There are some small things to clean up, but I like this PR. @dcbaker's support for muon is appreciated, but the build environment I'm most interested in doesn't have muon in the base system image and it would be a very nontrivial dependency to add at this time.

Aside: We don't have the analogous ability to download and build dependencies from source as needed with this PR. That's probably fine. We can add that later if we miss it. If someone wants a hint on how that might be possible, check out the [Dependency Providers](https://cmake.org/cmake/help/latest/command/cmake_language.html#dependency-providers) feature, which I'm aware of but haven't had reason to use personally.

File: src/CMakeLists.txt
Comment (dcbaker) [reply src/CMakeLists.txt:36]: Those are the minimum supported versions we use. Meson does have those wired, and I would rather shift left, ie, cmake/meson should fail to configure if you try to use fmt 7 rather than failing at build time, or worse, compiling and then failing in some weird way at runtime.

With my packager hat on, this kind of information helps downstreams too, because ubuntu (for example) can use this information to help decide if they want to take an update, or if the versions they're putting together actually work, or whether they might want to patch cps-config because they want it work with a dependency version they ship we don't support.

File: src/CMakeLists.txt
Comment (dcbaker) [reply src/CMakeLists.txt:3]: Meson will create a library with the common scheme for each platform. On Unices this will be `libcps.[a,so,dylb]`, on windows it'll be `cps.[a,lib,dll]`. I'm not sure what CMake does here.

File: tests/CMakeLists.txt
Comment (lunacd) [reply tests/CMakeLists.txt]: Looking at the commit history, I think this repo has been using rebasing to incorporate changes. So I'll try to keep the commit history clean on the PR by rebasing this change onto a previous commit

File: CMakeLists.txt
Comment (lunacd) [reply CMakeLists.txt:3]: Let's circle back to this on another PR

File: .gitignore
Comment (lunacd) [reply .gitignore:7]: What would be an alternative to ignoring that directory? Does cmake has an option to ignore its own out of source build directory?

File: src/CMakeLists.txt
Comment (lunacd) [reply src/CMakeLists.txt]: I only care that the cmake and meson does similar things. I'd be happy to either keep version check in both or neither.

File: src/cps/config.hpp.in
Comment (lunacd) [reply src/cps/config.hpp.in:3]: Making an issue to follow up on this

File: src/CMakeLists.txt
Comment (bretbrownjr) [reply src/CMakeLists.txt]: If you name the library `cps`, CMake will make a `libcps.a` or `libcps.so`, depending on whether you selected a static or shared build. The default is static, but someone building a project can choose shared with `-DBUILD_SHARED_LIB=ON`.

The big thing is to ship a library named `libcps.a`, yes. We could add a little validation to the CI if that's important to get right. Eventually, if we add support for Conan, there are ways to add a "packaging test" that includes linking against packaged libraries, which will be a more end-to-end way to ensure the same effect.

File: src/CMakeLists.txt
Comment (bretbrownjr) [reply src/CMakeLists.txt]: Well, the dependency management system (which could be CMake but specifically in FetchContent logic), is more "left" than the build step, and it's the part of the development process responsible for finding versions of dependencies that match, broadly.

Not overspecifying in the build system makes it easy to perform development activities at that layer. One doesn't have to fiddle with CMake and Meson in order to upgrade or downgrade dependencies, especially as a bulk operation.

File: .gitignore
Comment (lunacd) [reply .gitignore:7]: Oh I didn't realize I could define user-wide ignores. Then I'm guessing the `binaryDir` setting in `CMakePresets.json` also wouldn't be necessary then?

File: src/CMakeLists.txt
Comment (dcbaker) [reply src/CMakeLists.txt]: Maybe I'm missing something, but specifying versions as `== 1.2.1` is bad practice, 100% agreed. But saying `>= 1.2` when you in fact rely on features in 1.2 is good practice, since it helps users building your software to get things configured correctly immediately which lowers useless bug reports, and it documents for vendors what you expect. If a software vendor wants to use 1.1 for some reason (lets say Ubuntu in this case), they have mechanisms in place to carry the patch to Meson or CMake that does that as part of their packaging process.

Comment (lunacd) [reply src/CMakeLists.txt]: If I understand both of you correctly, I think we can agree a minimum version requirement that is not over-specified is beneficial. An over-specified version requirement is not as great. The question at hand then is: are we over-specifying? Does using a lower version of the library actually break anything?

Comment (dcbaker) [reply src/CMakeLists.txt]: In the case of fmt I think it does. I had to do work to make fmt < 9 work (which is why my distro was shipping), and so 
I'd be shocked if 6 or 7 worked as is. I think CLI11 also had issues with 2.0, which is why we have 2.1 there, but I might be misremembering. I know that the library we were using before CLI11 definitely had issues unless you were using specific versions.

**Comment by lunacd** [comment]
> Aside: We don't have the analogous ability to download and build dependencies from source as needed with this PR. That's probably fine. We can add that later if we miss it. If someone wants a hint on how that might be possible, check out the [Dependency Providers](https://cmake.org/cmake/help/latest/command/cmake_language.html#dependency-providers) feature, which I'm aware of but haven't had reason to use personally.

I have no idea this exists. Looks interesting so I took a shot at it. See the very last commit of this series. Let me know if I'm using something the wrong way.

One thing that confuses me is jsoncpp's cmake target name. Their namespaced target name is, quite surprisingly to me, provided in [a separate cmake file](https://github.com/open-source-parsers/jsoncpp/blob/69098a18b9af0c47549d9a271c054d13ca92b006/jsoncpp-namespaced-targets.cmake). The consequence is that when jsoncpp is downloaded and built from source, cmake does not see the namespaced target. Any idiomatic way to make it available?

And what confuses me even further is the two non-namespaced target it provides: `jsoncpp_static` and `jsoncpp_lib`. What's the difference between lib and static? lib sounds static... Anyway I used `jsoncpp_lib` in this commit because `static` doesn't link 🤷

**Comment by bretbrownjr** [comment]
> Any idiomatic way to make it available?

As far as I know, not especially. One of the things I'm hoping CPS can help with is for projects to have declarative ways (i.e., CPS files!) to explicitly define their interfaces.

Barring that, some procedural code detecting that the `JsonCpp::JsonCpp` is not defined and defining it with `add_library(ALIAS ...)` as a workaround we could live with for now.

As to the static versus shared question, looks like `jsoncpp_lib` [is a shared library](https://github.com/open-source-parsers/jsoncpp/blob/69098a18b9af0c47549d9a271c054d13ca92b006/src/lib_json/CMakeLists.txt#L116-L117). *Probably* we should just match the library type of the cps library target.

**Comment by lunacd** [comment]
> As to the static versus shared question, looks like jsoncpp_lib [is a shared library](https://github.com/open-source-parsers/jsoncpp/blob/69098a18b9af0c47549d9a271c054d13ca92b006/src/lib_json/CMakeLists.txt#L116-L117). Probably we should just match the library type of the cps library target.

Makes sense to me. I made it link against lib or static depending on `BUILD_SHARED_LIBS`

File: src/CMakeLists.txt
Comment (lunacd) [reply src/CMakeLists.txt]: I believe this remains as the only unresolved thread. @bretbrownjr @dcbaker What do you guys think?

Comment (bretbrownjr) [reply src/CMakeLists.txt]: Let's leave the version qualifiers in for now so we can move ahead with CMake support. I'm still concerned that they're overspecified inside `CMakeLists.txt`, especially since analogous information will be in the configurations for meson and relevant dependency management systems, but maybe we can follow up on that later.

**Review by bretbrownjr — APPROVED** [review]
I'm ready to move ahead on this, once the CI is working.

@lunacd is this related to this PR or did the CI in the main branch start failing somehow?

**Comment by lunacd** [comment]
```
error: failed to list workers: Unavailable: connection error: desc = "transport: Error while dialing: dial unix /run/buildkit/buildkitd.sock: connect: no such file or directory"
```
Looks like system failures on GitHub Actions side. Retrying

**Comment by lunacd** [comment]
Can't figure out how to on the UI. I guess I'll just force push tonight

**Comment by lunacd** [comment]
Looks like a docker layer caching problem. I pushed another commit to fix that. This probably doesn't belong here but it wouldn't pass without. So just don't squash the commits. Merge or rebase instead

**Pull request merged**
